core/lease/storedb.go

```diff
@@ -0,0 +1,365 @@
+
+		result, err := s.db.ExecContext(
+			ctx, q, d, lease.Namespace, lease.ModelUUID, lease.Lease, request.Holder)
+
+		// If no rows were affected, then either this lease does not exist or
+		// it is not held by the input holder, constituting an invalid request.
+		if err == nil {
+			var affected int64
+			affected, err = result.RowsAffected()
+			if affected == 0 && err == nil {
+				err = ErrInvalid
+			}
+		}
+		errCh <- err
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		return errors.New("extend cancelled")
+	case err := <-errCh:
+		cancel()
+		return errors.Trace(err)
+	}
+}
+
+// RevokeLease (lease.Store) deletes the lease from the store,
+// provided it exists and is held by the input holder.
+// If either of these conditions is false, an error is returned.
+func (s *DBStore) RevokeLease(lease Key, holder string, stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+DELETE FROM lease
+WHERE  uuid = (
+    SELECT l.uuid
+    FROM   lease l JOIN lease_type t ON l.lease_type_id = t.id
+    WHERE  t.type = ?
+    AND    l.model_uuid = ?
+    AND    l.name = ?
+    AND    l.holder = ?
+)`[1:]
+
+		result, err := s.db.ExecContext(
+			ctx, q, lease.Namespace, lease.ModelUUID, lease.Lease, holder)
+
+		if err == nil {
+			var affected int64
+			affected, err = result.RowsAffected()
+			if affected == 0 && err == nil {
+				err = ErrInvalid
+			}
+		}
+		errCh <- err
+	}()
+
+	select {
+	case <-stop:
+		cancel()
+		return errors.New("revoke cancelled")
+	case err := <-errCh:
+		cancel()
+		return errors.Trace(err)
+	}
+}
+
+// LeaseGroup (lease.Store) returns all leases
+// for the input namespace and model.
+func (s *DBStore) LeaseGroup(namespace, modelUUID string) (map[Key]Info, error) {
+	q := `
+SELECT t.type, l.model_uuid, l.name, l.holder, l.expiry
+FROM   lease l JOIN lease_type t ON l.lease_type_id = t.id
+WHERE  t.type = ?
+AND    l.model_uuid = ?`[1:]
+
+	rows, err := s.db.Query(q, namespace, modelUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	result, err := leasesFromRows(rows)
+	return result, errors.Trace(err)
+}
+
+// PinLease (lease.Store) adds the input entity into the lease_pin table
+// to indicate that the lease indicated by the input key must not expire,
+// and that this entity requires such behaviour.
+func (s *DBStore) PinLease(lease Key, entity string, stop <-chan struct{}) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	errCh := make(chan error)
+
+	go func() {
+		q := `
+INSERT INTO lease_pin (uuid, lease_uuid, entity_id)
+SELECT ?, l.uuid, ?
+FROM   lease l JOIN lease_type t ON l.lease_type_id = t.id
+WHERE  t.type = ?
+AND    l.model_uuid = ?
+AND    l.name = ?`[1:]
+
+		_, err := s.db.ExecContext(
+			ctx, q, utils.MustNewUUID().String(), entity, lease.Namespace, lease.ModelUUID, lease.Lease)
+
+		errCh <- err
+	}()
+
+	select {
+	case <-stop:
+		cancel()
```

Comment:
I'm curious about this request response design. If we do trigger cancel(), shouldn't we still wait on <-err to see that the request has actually been stopped and the goroutine returned?

Reply:
I was always calling `cancel`, but we wait after cancellation everywhere with:
```go
case <-stop:
    cancel()
    return errors.Trace(<-errCh)
```